modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/introspection/validation/NameClashModelValidator.java

```diff
@@ -213,6 +258,68 @@ private void validateClash(String existingNamingModel, String newNamingModel, St
                                                          typeOfNewNamingModel));
       }
     }
+
+    private void validateSingularizingClash(ParameterizedModel model) {
+      List<ParameterModel> parameters = model.getParameterModels();
+      parameters.forEach(
+                         p -> p.getType().accept(new MetadataTypeVisitor() {
+
+                           @Override
+                           public void visitArrayType(ArrayType arrayType) {
+                             validateSingular();
+                           }
+
+                           @Override
+                           public void visitDictionary(DictionaryType dictionaryType) {
+                             validateSingular();
+                           }
+
+                           private void validateSingular() {
+                             String singularName = singularize(p.getName());
+                             boolean allowInline = p.getModelProperty(XmlHintsModelProperty.class)
+                                 .map(XmlHintsModelProperty::allowsInlineDefinition)
+                                 .orElse(true);
+
+                             if (singularName.equals(p.getName()) || !allowInline) {
+                               return;
+                             }
+
+                             Optional<ParameterModel> clashingParam =
+                                 parameters.stream().filter(p -> p.getName().equals(singularName)).findAny();
+                             String describedReference = new DescribedReference<>(model).getDescription();
+                             Class<Object> type = getType(p.getType());
+                             String extensionName = extensionModel.getName();
+
+                             if (clashingParam.isPresent()) {
+                               throw new IllegalModelDefinitionException(format("Extension '%s' defines an %s of name '%s' which contains a '%s' "
+                                   + "parameter '%s' that when singularized clashes with another "
+                                   + "parameter '%s' in the same %s", extensionName,
+                                                                                describedReference, model.getName(),
+                                                                                type.getSimpleName(),
+                                                                                p.getName(), clashingParam.get().getName(),
+                                                                                describedReference));
+                             }
+
+                             Optional<String> nameClash =
```

Comment:
all this, including the exception throwing => new method
validateNameClashWithTopLevelElements?